geode-protobuf-client/src/main/java/org/apache/geode/internal/cache/client/protobuf/Region.java

```diff
@@ -0,0 +1,204 @@
+        .setValue(encodeValue(unencodedValue)).build();
+  }
+
+  Region(String name, Socket socket) {
+    this.name = name;
+    this.socket = socket;
+  }
+
+  public V get(K key) throws Exception {
+    final OutputStream outputStream = socket.getOutputStream();
+    ClientProtocol.Message.newBuilder()
+        .setRequest(ClientProtocol.Request.newBuilder().setGetRequest(RegionAPI.GetRequest
+            .newBuilder().setRegionName(name).setKey(encodeValue(key.toString()))))
+        .build().writeDelimitedTo(outputStream);
+
+    // TODO: How does one get a java.lang.Object out of
+    // org.apache.geode.internal.protocol.protobuf.v1.EncodedValue?
+    final InputStream inputStream = socket.getInputStream();
+    return (V) ClientProtocol.Message.parseDelimitedFrom(inputStream).getResponse().getGetResponse()
+        .getResult().getStringResult();
+  }
+
+  public Map<K, V> getAll(Collection<K> keys) throws Exception {
+    Map<K, V> values = new HashMap<>();
+
+    final OutputStream outputStream = socket.getOutputStream();
+    RegionAPI.GetAllRequest.Builder getAllRequest = RegionAPI.GetAllRequest.newBuilder();
+    getAllRequest.setRegionName(name);
+    for (K key : keys) {
+      getAllRequest.addKey(encodeValue(key.toString()));
+    }
+    ClientProtocol.Message.newBuilder()
+        .setRequest(ClientProtocol.Request.newBuilder().setGetAllRequest(getAllRequest)).build()
+        .writeDelimitedTo(outputStream);
+
+    final InputStream inputStream = socket.getInputStream();
+    final RegionAPI.GetAllResponse getAllResponse =
+        ClientProtocol.Message.parseDelimitedFrom(inputStream).getResponse().getGetAllResponse();
+    for (BasicTypes.Entry entry : getAllResponse.getEntriesList()) {
+      // TODO: How does one get a java.lang.Object out of
+      // org.apache.geode.internal.protocol.protobuf.v1.EncodedValue?
+      // TODO values.put((K) entry.getKey().getStringResult(), (V)
+      // entry.getValue().getStringResult());
+    }
+
+    return values;
+  }
+
+  public void put(K key, V value) throws Exception {
+    final OutputStream outputStream = socket.getOutputStream();
+    ClientProtocol.Message.newBuilder()
+        .setRequest(ClientProtocol.Request.newBuilder()
+            .setPutRequest(RegionAPI.PutRequest.newBuilder().setRegionName(name)
+                .setEntry(encodeEntry(key.toString(), value))))
+        .build().writeDelimitedTo(outputStream);
+
+    final InputStream inputStream = socket.getInputStream();
+    ClientProtocol.Message.parseDelimitedFrom(inputStream).getResponse().getPutResponse();
+  }
+
+  public void putAll(Map<K, V> values) throws Exception {
+    final OutputStream outputStream = socket.getOutputStream();
+    RegionAPI.PutAllRequest.Builder putAllRequest = RegionAPI.PutAllRequest.newBuilder();
+    putAllRequest.setRegionName(name);
+    for (K key : values.keySet()) {
+      putAllRequest.addEntry(encodeEntry(key, values.get(key)));
```

Comment:
You can avoid all these values.get calls by:
for (Map.Entry<K,V> entry : values.entrySet()) { putAllRequest.addEntry(encodeEntry(entry.getKey(), entry.getValue)); }